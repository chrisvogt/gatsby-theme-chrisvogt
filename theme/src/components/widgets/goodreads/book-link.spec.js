import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import { navigate as gatsbyNavigate } from 'gatsby'

import BookLink from './book-link'

// Mock gatsby's navigate function
jest.mock('gatsby', () => ({
  navigate: jest.fn()
}))

describe('Widget/Goodreads/BookLink', () => {
  const mockProps = {
    id: '123',
    title: 'Test Book',
    thumbnailURL: 'https://example.com/book.jpg'
  }

  beforeEach(() => {
    jest.clearAllMocks()
    // Mock console.log to keep test output clean
    jest.spyOn(console, 'log').mockImplementation(() => {})
  })

  it('renders a book link with the correct attributes', () => {
    render(<BookLink {...mockProps} />)
    const link = screen.getByTestId('book-link')
    expect(link).toHaveAttribute('href', '?bookId=123')
    expect(link).toHaveAttribute('title', 'Test Book')
  })

  it('handles CDN URLs by appending webp format', () => {
    const cdnProps = {
      ...mockProps,
      thumbnailURL: 'https://images.imgix.net/book.jpg'
    }
    render(<BookLink {...cdnProps} />)
    const image = screen.getByTestId('book-preview-thumbnail')
    expect(image).toHaveAttribute('xlink:href', 'https://images.imgix.net/book.jpg?auto=compress&auto=format')
  })

  it('preserves non-CDN URLs without modification', () => {
    render(<BookLink {...mockProps} />)
    const image = screen.getByTestId('book-preview-thumbnail')
    expect(image).toHaveAttribute('xlink:href', 'https://example.com/book.jpg')
  })

  it('handles click events with scroll position preservation', async () => {
    // Mock window.scrollY
    Object.defineProperty(window, 'scrollY', { value: 200 })
    render(<BookLink {...mockProps} />)
    const link = screen.getByTestId('book-link')
    fireEvent.click(link)
    // Wait for the setTimeout to complete
    await new Promise(resolve => setTimeout(resolve, 0))
    expect(gatsbyNavigate).toHaveBeenCalledWith('?bookId=123', {
      replace: true,
      state: {
        noScroll: true,
        scrollPosition: 200
      }
    })
  })
})
