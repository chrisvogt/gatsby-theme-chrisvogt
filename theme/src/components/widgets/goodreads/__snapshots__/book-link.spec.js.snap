// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookLink matches the snapshot 1`] = `
<a
  className="css-1avyp1d"
  href="https://www.google.com/books/edition/_/ZrNzAwAAQBAJ?hl=en"
  title="The Three-Body Problem"
>
  <svg
    className="css-kklm2g-Book"
    data-name="Book Preview"
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
    xmlnsXlink="http://www.w3.org/1999/xlink"
  >
    <defs>
      <clipPath
        id="ec3c1a72-0104-4d8a-830f-57d2245cb2f4"
      >
        <polyline
          data-name="Book Cover"
          fill="none"
          points="78.6 38.6 78.6 88.3 14.9 88.3 14.9 0 78.6 0 78.6 43.6"
        />
      </clipPath>
    </defs>
    <title>
      The Three-Body Problem on Google Books
    </title>
    <g
      data-name="Book Sheets"
      fill="#1a202c"
      id="bc5c39bc-971a-4562-b8d4-0c72a22c3f5f"
    >
      <path
        d="M78.6,90.9H12.3V0H78.6ZM14.9,88.3H76V2.6H14.9Z"
      />
      <rect
        height="46.2"
        width="2.61"
        x="85.1"
        y="9.1"
      />
      <polygon
        points="87.7 100 20.83 100 12.35 90.89 14.25 89.11 21.97 97.4 85.1 97.4 85.1 49 87.7 49 87.7 100"
      />
      <polygon
        points="80.06 92.36 18.8 92.36 18.8 92.04 79.74 92.04 79.74 3.4 80.06 3.4 80.06 92.36"
      />
      <polygon
        points="81.36 93.66 20.1 93.66 20.1 93.34 81.04 93.34 81.04 4.7 81.36 4.7 81.36 93.66"
      />
      <polygon
        points="82.66 94.96 21.4 94.96 21.4 94.64 82.34 94.64 82.34 6 82.66 6 82.66 94.96"
      />
      <polygon
        points="83.96 96.26 22.7 96.26 22.7 95.94 83.64 95.94 83.64 7.3 83.96 7.3 83.96 96.26"
      />
    </g>
    <polyline
      data-name="Image Wrapper"
      fill="#1a202c"
      points="77.3 39.9 77.3 89.6 13.6 89.6 13.6 1.3 77.3 1.3 77.3 44.9"
    />
    <g
      style={
        {
          "clipPath": "url(#ec3c1a72-0104-4d8a-830f-57d2245cb2f4)",
        }
      }
    >
      <image
        data-testid="book-preview-thumbnail"
        height="190"
        transform="translate(14.53 -3.68) scale(0.5)"
        width="128"
        xlinkHref="https://placehold.it/400/400"
      />
    </g>
  </svg>
</a>
`;
